src/include/migraphx/find_concur.hpp

```diff
@@ -0,0 +1,243 @@
+#ifndef MIGRAPHX_GUARD_FIND_CONCUR_HPP
+#define MIGRAPHX_GUARD_FIND_CONCUR_HPP
+
+#include <cassert>
+#include <string>
+#include <functional>
+#include <memory>
+#include <type_traits>
+#include <utility>
+#include <unordered_map>
+#include <vector>
+
+#include <migraphx/instruction.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+struct program;
+
+#ifdef DOXYGEN
+
+/// An interface for target-dependent analysis to find concurrent instructions
+/// executing in different streams.
+struct find_concur
+{
+    void get_concur(program* p,
+                    int num_of_streams,
+                    std::unordered_map<const instruction*,
+                                       std::vector<std::vector<const instruction*>>>& concur_instrs,
+                    std::unordered_map<const instruction*, int>& instr2_points);
```

Comment:
This should be the return value not an input parameter.

Reply:
Return value is passed by reference.  Memory is allocated from caller.
Returning a data structure involves two copies, first copying the data allocated from callee into temporaries, then copy the temporaries into data allocated from caller.

Reply:
> Return value is passed by reference.

But this should not be extending the data structure. If the user passes the map with previous values it can produce different results. 

>  Memory is allocated from caller.

Memory is allocated by `unordered_map`.

> Returning a data structure involves two copies, first copying the data allocated from callee into temporaries, then copy the temporaries into data allocated from caller.

NRVO+[guaranteed copy elision](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/p0135r0.html) will get rid of the intermediate copies.

In addition, taking the parameter by reference creates aliasing issues that can prevent optimizations. So doing it this way is a ppessimization.

But more importantly, returning the "return value" makes the semantics of the function much clearer.

Reply:
There is no guarantee that compiler can get rid of temporary copies for return-by-value objects.
It is also a calling convention requirement.

Return a value whose memory is allocated in the callee is not a good idea.

Reply:
x86 calling convention for return-by-value:  compiler calls allocate to creates a temporary storage from caller's frame, then pass it by reference to callee.

Reply:
> There is no guarantee that compiler can get rid of temporary copies for return-by-value objects.

Yes there is. That is what P0135 guarantees. That is why you can return objects that are not copyable or movable:

```cpp
struct NonMoveable {
  NonMoveable(int);
  NonMoveable(NonMoveable&) = delete;
  void NonMoveable(NonMoveable&) = delete;
  std::array<int, 1024> arr;
};
NonMoveable make() {
  return NonMoveable(42); // ok, directly constructs returned object
}
auto nm = make(); // ok, directly constructs 'nm'
```

Either way, if you dont want to change it, then leave it, and I will update it later.